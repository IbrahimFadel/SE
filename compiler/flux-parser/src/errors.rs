use ariadne::Report;
use flux_lexer::TokenKind;
use flux_span::{Span, Spanned};

#[derive(Debug, Clone, PartialEq)]
pub enum ParseError {
	ExpectedToken {
		expected: Vec<TokenKind>,
		got: Spanned<TokenKind>,
	},
	// More general which can allow for more specific messages and for expecting larger constructs like types or parameter
	Expected {
		expected: Spanned<String>,
	},
}

impl<'a> Into<&'a Report<Span>> for &'a ParseError {
	fn into(self) -> &'a Report<Span> {
		todo!()
	}
}
