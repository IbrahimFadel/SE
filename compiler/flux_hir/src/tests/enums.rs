use crate::errors;

errors! {
    unuses_generic_params,
    r#"
//- main.flx

enum Result<A, B, C> {
    A -> A,
    B,
    C -> C
}
"#
}

errors! {
    type_mismatch,
    r#"
//- main.flx

enum Foo {
    A,
    B -> u32,
}

fn main() {
    let x u32 = Foo::A;
    let x u32 = Foo::B(0);
}
"#
}

errors! {
    missing_arg,
    r#"
//- main.flx

enum Foo {
    A -> s32,
}

fn main() {
    let x = Foo::A;
}
"#
}

errors! {
    incorrect_num_args,
    r#"
//- main.flx

enum Foo {
    A -> s32,
}

fn main() {
    let x = Foo::A(0, 0);
}
"#
}

errors! {
    type_mismatch_in_args,
    r#"
//- main.flx

enum Foo {
    A -> s32,
}

fn main() {
    let x = Foo::A(0.0);
}
"#
}
