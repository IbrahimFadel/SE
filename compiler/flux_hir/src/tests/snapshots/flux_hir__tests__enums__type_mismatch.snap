---
source: compiler/flux_hir/src/tests/enums.rs
expression: s
---
//  main.flx


fn main() -> () {
    let x Foo = Foo::A;
    let x Foo = Foo::B(0)(0);
}

Diagnostics

[E0039] Error: type mismatch
   ,-[ main.flx:9:9]
   |
 9 |     let x u32 = Foo::A;
   |   ^^|^  ^^^|^^  
   |     `----------- `u32`
   |            |    
   |            `---- type mismatch between `u32` and `Foo`
   |            |    
   |            `---- `Foo`
---'
[E0039] Error: type mismatch
    ,-[ main.flx:10:8]
    |
 10 |     let x u32 = Foo::B(0);
    |  ^^|^  ^^^||^^^^  
    |    `-------------- `u32`
    |           ||      
    |           `------- `Foo`
    |            |      
    |            `------ type mismatch between `u32` and `Foo`
----'

